### PR TITLE
Add git checks to npm release

### PR DESCRIPTION
**Pull request by levithomason**
This PR adds a check to ensure the working tree is clean.  It also ensures you are on the latest version of the branch.  This prevents failures when releasing due to needing to push/pull changes.

Lastly, the script will now fail at any of the check points instead of powering through when it is obvious the release cannot continue.

### CODE REVIEW

File: npm/release.sh [hunk 1]
```diff
@@ -12,28 +13,68 @@ usage() {
   echo "  Usage: bash $0 <major|minor|patch>"
 }
 
+print() {
+  echo "NPM RELEASE: $1"
+}
+
```
Comment (levithomason) [inline npm/release.sh]: Added a print util as well for more readable output.

File: npm/release.sh [hunk 2]
```diff
@@ -12,28 +13,68 @@ usage() {
...
   local version=$1
+
+  # ensure git is ready
+  local local_sha=$(git rev-parse @)
+  local remote_sha=$(git rev-parse @{u})
+  local base_sha=$(git merge-base @ @{u})
+
+  if [[ -n $(git status --porcelain) ]]; then
+    print "Commit or stash you changes before releasing"
+    exit 1
```
Comment (levithomason) [inline npm/release.sh]: Added exit failures everywhere it made sense.

File: npm/release.sh [hunk 3]
```diff
@@ -12,28 +13,68 @@ usage() {
...
+  npm version ${version}
+
+  print "...git push"
+  git push
+
+  print "...git push --follow-tags"
+  git push --follow-tags
+
+  print "...npm publish"
+  npm publish
```
Comment (levithomason) [inline npm/release.sh]: Also printed headings before each command so it is more clear what is happening when you need to fix a bad release.

**Comment by davezuko** [comment]
👻 

You may be interested in reading: https://github.com/davidchambers/xyz

**Pull request merged**